x-pack/plugins/infra/public/containers/saved_view/saved_view.tsx

```diff
@@ -0,0 +1,262 @@
+import { useSourceConfigurationFormState } from '../../components/source_configuration/source_configuration_form_state';
+import { useGetSavedObject } from '../../hooks/use_get_saved_object';
+import { useUpdateSavedObject } from '../../hooks/use_update_saved_object';
+
+export type SavedView<ViewState> = ViewState & {
+  name: string;
+  id: string;
+  isDefault?: boolean;
+};
+
+export type SavedViewSavedObject<ViewState = {}> = ViewState & {
+  name: string;
+};
+
+export type ViewType =
+  | typeof metricsExplorerViewSavedObjectName
+  | typeof inventoryViewSavedObjectName;
+
+interface Props {
+  defaultViewState: SavedView<any>;
+  viewType: ViewType;
+  shouldLoadDefault: boolean;
+}
+
+export const useSavedView = (props: Props) => {
+  const {
+    source,
+    isLoading: sourceIsLoading,
+    sourceExists,
+    createSourceConfiguration,
+    updateSourceConfiguration,
+  } = useContext(Source.Context);
+  const { viewType, defaultViewState } = props;
+  type ViewState = typeof defaultViewState;
+  const { data, loading, find, error: errorOnFind, hasView } = useFindSavedObject<
+    SavedViewSavedObject<ViewState>
+  >(viewType);
+
+  const [currentView, setCurrentView] = useState<SavedView<any> | null>(null);
+  const [loadingDefaultView, setLoadingDefaultView] = useState<boolean | null>(null);
+  const { create, error: errorOnCreate, data: createdViewData, createdId } = useCreateSavedObject(
+    viewType
+  );
+  const { update, error: errorOnUpdate, data: updatedViewData, updatedId } = useUpdateSavedObject(
+    viewType
+  );
+  const { deleteObject, deletedId } = useDeleteSavedObject(viewType);
+  const { getObject, data: currentViewSavedObject } = useGetSavedObject(viewType);
+  const [createError, setCreateError] = useState<string | null>(null);
+
+  useEffect(() => setCreateError(errorOnCreate), [errorOnCreate]);
+
+  const deleteView = useCallback((id: string) => deleteObject(id), [deleteObject]);
+  const formState = useSourceConfigurationFormState(source && source.configuration);
+  const defaultViewFieldName = useMemo(
+    () => (viewType === 'inventory-view' ? 'inventoryDefaultView' : 'metricsExplorerDefaultView'),
+    [viewType]
+  );
+
+  const makeDefault = useCallback(
+    async (id: string) => {
+      if (sourceExists) {
+        await updateSourceConfiguration({
+          ...formState.formStateChanges,
+          [defaultViewFieldName]: id,
+        });
+      } else {
+        await createSourceConfiguration({
+          ...formState.formState,
+          [defaultViewFieldName]: id,
+        });
+      }
+    },
+    [
+      formState.formState,
+      formState.formStateChanges,
+      sourceExists,
+      defaultViewFieldName,
+      createSourceConfiguration,
+      updateSourceConfiguration,
+    ]
+  );
+
+  const saveView = useCallback(
+    (d: { [p: string]: any }) => {
+      const doSave = async () => {
+        const exists = await hasView(d.name);
```

Comment:
not really, the logic is slightly different. For update, it's fine if another view exists with the same name, but not for create. I think it's probably fine since it's really only a single if.